custom-ktlint-rules/src/main/kotlin/org/partiql/ktlint/rule/TopLevelInternalRule.kt

```diff
@@ -0,0 +1,38 @@
+package org.partiql.ktlint.rule
+
+import com.pinterest.ktlint.core.Rule
+import com.pinterest.ktlint.core.ast.ElementType
+import com.pinterest.ktlint.core.ast.children
+import org.jetbrains.kotlin.com.intellij.lang.ASTNode
+
+public class TopLevelInternalRule : Rule("top-level-internal") {
+
+    override fun visit(
+        node: ASTNode,
+        autoCorrect: Boolean,
+        emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit
+    ) {
+        if (node.elementType != ElementType.IDENTIFIER) {
+            return
+        }
+
+        // Focus on just functions and values
+        val parent = node.treeParent
+        if (parent.elementType != ElementType.FUN && parent.elementType != ElementType.PROPERTY) {
+            return
+        }
+
+        // Check grandparent of node is FILE; if so, is top-level declaration
+        if (parent.treeParent.elementType != ElementType.FILE) {
+            return
+        }
+        val modifiers = parent.findChildByType(ElementType.MODIFIER_LIST)?.children()
+        if (modifiers != null && modifiers.any { it.elementType == ElementType.INTERNAL_KEYWORD }) {
```

Comment:
How about top-level modifiers *required* `public` otherwise fail

Reply:
We could have `private` top-level declarations, which won't be visible from Java. So I think just finding `internal` declarations is suitable enough for this ktlint rule.

Requiring that there is a modifier should be caught by enabling explicit API mode.